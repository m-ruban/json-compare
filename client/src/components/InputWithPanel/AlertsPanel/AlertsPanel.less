.alerts {
  height: 40%;

  &__items {
    padding: 0.375rem 0.75rem;
    border-radius: 0.25rem;
    height: 100%;
    border: 1px solid #ced4da;
    background-color: #f8f9fa;
    font-size: 0.8rem;
    overflow-x: hidden;
    overflow-y: auto;
  }
}
