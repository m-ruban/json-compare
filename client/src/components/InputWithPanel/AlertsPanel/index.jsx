import React from 'react';
import PropTypes from 'prop-types';
import { Col, Row, Badge } from 'reactstrap';

import withConsumer from 'hoc/withConsumer';
import trl from 'modules/translation';
import { COMPARE_REQ } from 'constants/compare';

const AlertsPanel = ({ alerts }) => {
    const nodesAlerts = [],
        nodesRequired = [];

    alerts.forEach((alert, path) => {
        if (!alert.scalar) return;
        if (alert.res === COMPARE_REQ) {
            nodesRequired.push(<div key={path}>{path}</div>);
        } else {
            nodesAlerts.push(<div key={path}>{path}</div>);
        }
    });

    return (
        <div className="pt-4 panel">
            <div className="body">
                <Row>
                    <Col xs="6">
                        <Badge color="diff" pill>
                            !
                        </Badge>
                        &nbsp;
                        <b>{trl('ResultPanel.alerts')}</b>
                        <br />
                        {nodesAlerts}
                    </Col>
                    <Col xs="6">
                        <Badge color="req" pill>
                            !
                        </Badge>
                        &nbsp;
                        <b>{trl('ResultPanel.required')}</b>
                        <br />
                        {nodesRequired}
                    </Col>
                </Row>
            </div>
        </div>
    );
};

AlertsPanel.propTypes = {
    alerts: PropTypes.object,
};

export default withConsumer(AlertsPanel);
