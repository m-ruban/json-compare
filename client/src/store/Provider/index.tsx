import * as React from 'react';

import { COMPARE_REQ } from 'constants/compare';
import { EQUALITY_TYPE_STRICT } from 'constants/equality';
import { AUTHOR_LINK, FEEDBACK_LINK } from 'constants/links';
import { LOG_INFO } from 'constants/log';
import { Provider } from 'store/Context';

import { equalAlerts } from 'modules/compare';
import { parseAndLog } from 'modules/parse';

interface IProviderProps {
    children: any;
}

interface IProviderState {
    alerts: Map<string, object>;
    author: string;
    changeEqualityTypes: (event: React.SyntheticEvent) => void;
    changeObj: (event: React.SyntheticEvent) => void;
    changeTool: (value: string) => void;
    equalityTypes: string;
    feedback: string;
    isCompare: boolean;
    log: object;
    objObject: object;
    objStr: string;
    toggleCompare: (event: React.SyntheticEvent) => void;
    toolObject: object;
    toolStr: string;
}

class AppProvider extends React.Component<IProviderProps, IProviderState> {
    constructor(props) {
        super(props);
        this.state = {
            alerts: new Map(),
            author: AUTHOR_LINK,
            changeEqualityTypes: this.changeEqualityTypes.bind(this),
            changeObj: this.changeObj.bind(this),
            changeTool: this.changeTool.bind(this),
            equalityTypes: EQUALITY_TYPE_STRICT,
            feedback: FEEDBACK_LINK,
            isCompare: false,
            log: { text: '', type: null },
            objObject: {},
            objStr: '',
            toggleCompare: this.toggleCompare.bind(this),
            toolObject: {},
            toolStr: '',
        };
    }

    public toggleCompare() {
        const log = {
            text: 'toggle compare',
            type: LOG_INFO,
        };

        this.setState({ isCompare: !this.state.isCompare, log }, this.changeAlerts);
    }

    public changeEqualityTypes(event) {
        const { value } = event.target;
        const log = {
            text: 'change equality types',
            type: LOG_INFO,
        };
        this.setState({ equalityTypes: value, log }, this.changeAlerts);
    }

    public changeTool(value: string) {
        const { obj, log } = parseAndLog(value);
        this.setState({ toolObject: obj, toolStr: value, log }, this.changeAlerts);
    }

    public changeObj(value: string) {
        const { obj, log } = parseAndLog(value);
        this.setState({ objObject: obj, objStr: value, log }, this.changeAlerts);
    }

    public changeAlerts() {
        const { toolObject, objObject, equalityTypes, isCompare } = this.state;
        const isEmpty = (obj) => Object.keys(obj).length === 0;

        if (!isEmpty(toolObject) && !isEmpty(objObject) && isCompare) {
            const alerts = equalAlerts(toolObject, objObject, equalityTypes);
            let req = 0;
            let diff = 0;
            let text = '';

            alerts.forEach((alert) => {
                if (!alert.scalar) { return; }
                if (alert.res === COMPARE_REQ) {
                    req++;
                } else {
                    diff++;
                }
            });

            if (diff > 0 || req > 0) {
                text = `${diff} alert${diff > 1 ? 's' : ''}, ${req} required`;
            } else {
                text = 'objects are equal';
            }

            this.setState({
                alerts,
                log: { text, type: LOG_INFO },
            });
        }
    }

    public render() {
        return <Provider value={this.state}>{this.props.children}</Provider>;
    }
}

export default AppProvider;
