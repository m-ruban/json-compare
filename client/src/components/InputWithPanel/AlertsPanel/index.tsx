import React from 'react';
import { Col, Row } from 'reactstrap';

import Mark from 'components/Mark';
import { COMPARE_REQ } from 'constants/compare';
import withConsumer from 'hoc/withConsumer';
import trl from 'modules/translation';

import 'components/InputWithPanel/AlertsPanel/AlertsPanel.less';

interface IAlert {
    res: string;
    scalar: boolean;
}

interface IAlertsPanelProps {
    alerts: Map<string, IAlert>;
}

const AlertsPanel = ({ alerts }: IAlertsPanelProps) => {
    const nodesAlerts = [];
    const nodesRequired = [];

    alerts.forEach((alert, path) => {
        if (!alert.scalar) {
            return;
        }
        if (alert.res === COMPARE_REQ) {
            nodesRequired.push(<div key={path}>{path}</div>);
        } else {
            nodesAlerts.push(<div key={path}>{path}</div>);
        }
    });

    return (
        <div className="pt-4 alerts">
            <div className="alerts__items bg-dark">
                <Row>
                    <Col xs="6">
                        <Mark type="diff" text="!" pill />
                        &nbsp;
                        <b>{trl('ResultPanel.alerts')}</b>
                        <br />
                        {nodesAlerts}
                    </Col>
                    <Col xs="6">
                        <Mark type="req" text="!" pill />
                        &nbsp;
                        <b>{trl('ResultPanel.required')}</b>
                        <br />
                        {nodesRequired}
                    </Col>
                </Row>
            </div>
        </div>
    );
};

export default withConsumer(AlertsPanel);
