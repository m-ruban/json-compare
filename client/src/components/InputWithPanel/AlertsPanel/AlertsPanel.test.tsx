import React from 'react';

import AlertsPanel from 'components/InputWithPanel/AlertsPanel';
import { render } from 'enzyme';
import { Provider} from 'store/Context';
import alertsExample from 'tests/mocks/alerts';

describe('<AlertsPanel />', () => {
    it('Renders correctly (pill false)', () => {
        const output = render(
            <Provider value={alertsExample}>
                <AlertsPanel />
            </Provider>,
        );
        expect(output).toMatchSnapshot();
    });
});
