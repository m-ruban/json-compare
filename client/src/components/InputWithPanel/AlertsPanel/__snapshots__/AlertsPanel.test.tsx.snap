// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AlertsPanel /> Renders correctly (pill false) 1`] = `
<div
  class="pt-4 alerts"
>
  <div
    class="alerts__items bg-dark"
  >
    <div
      class="row"
    >
      <div
        class="col-6"
      >
        <span
          class="json-mark json-mark_diff badge badge-json badge-pill"
        >
          !
        </span>
         
        <b>
          Alerts
        </b>
        <br />
        <div>
          test
        </div>
      </div>
      <div
        class="col-6"
      >
        <span
          class="json-mark json-mark_req badge badge-json badge-pill"
        >
          !
        </span>
         
        <b>
          Required
        </b>
        <br />
      </div>
    </div>
  </div>
</div>
`;
